[Core] Double (and single) Lagrange multiplier builder and solver

loumalouomega opened the pull request:
This adds a double Lagrange multiplier builder and solver (in order to solve MPC). It is based on the following work: https://www.code-aster.org/V2/doc/v14/en/man_r/r3/r3.03.01.pdf Original implementation is in Cast3m from CEA, and later ported and documented in CodeAster. 

Consider double LM allows to use iterative solvers, including AMGCL (with the cost of duplicate the number of LM, or unkowns). The LM also allows to consider non-linear constraints more easily.

This PR adds several tests and it depends on the following PRs: #5919 #5906 and #5905

loumalouomega committed:
Adding CONSTRAINT_SCALE_FACTOR variable

loumalouomega committed:
Minor

loumalouomega committed:
Minor

loumalouomega committed:
No transpose

loumalouomega committed:
More general

loumalouomega committed:
Adding first test

loumalouomega committed:
One more test

loumalouomega committed:
More tests

loumalouomega committed:
Adding to solver

loumalouomega committed:
Toward more advanced test

loumalouomega committed:
[skip ci] Merge master

loumalouomega committed:
Clean up

loumalouomega committed:
Additional check

loumalouomega committed:
Adding debug check

loumalouomega committed:
More info

loumalouomega committed:
Additional check

loumalouomega committed:
Minor

loumalouomega committed:
Minor

loumalouomega committed:
Better

loumalouomega committed:
WIP

loumalouomega committed:
Resize to avoid problems

loumalouomega committed:
More consistent call

loumalouomega committed:
Unused

loumalouomega committed:
Initialize mt if not initialized

loumalouomega committed:
Extending with basic test

loumalouomega committed:
Minor

loumalouomega committed:
Minor

loumalouomega committed:
Making method virtual

loumalouomega committed:
Override

loumalouomega committed:
More override

loumalouomega committed:
Using ublas space

loumalouomega committed:
Free memory

loumalouomega committed:
[Core] Consistent BuildRHSAndSolver with BuildAndSolve

The calls from BuildRHSAndSolver are not consistent with BuildAndSolve. Nor constraints ror Diritchlet conditions are applied

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Adding LM variable

loumalouomega committed:
Add ApplyRHSConstraints

loumalouomega committed:
Computing more terms

loumalouomega committed:
Missing resize

loumalouomega committed:
Activate test

loumalouomega committed:
Typo

loumalouomega committed:
Activate

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Reminder for AMGCL debug

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Update signs and stuff

loumalouomega committed:
Revert sign

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Update

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Activating test

loumalouomega committed:
Improve test

loumalouomega committed:
Reactivate tests

loumalouomega committed:
Relative check

loumalouomega committed:
Remove

loumalouomega committed:
Reduce number of iterations

loumalouomega committed:
Refactor assemble

loumalouomega committed:
Common filling

loumalouomega committed:
Missing pragma atomic

loumalouomega committed:
Transpose before (to avoid repeat the trasnpose)

loumalouomega committed:
Refactor

loumalouomega committed:
Deactivate test

loumalouomega committed:
Adding extended check in debug

loumalouomega committed:
Additional check

loumalouomega committed:
Assembling when is alive in memory

loumalouomega committed:
Update test

loumalouomega committed:
Minor

loumalouomega committed:
Missing memory release

loumalouomega committed:
Minor

loumalouomega committed:
Free memory

loumalouomega committed:
Fill with zeros

loumalouomega committed:
Activate missing tests

loumalouomega committed:
More realistic convergence criteria

loumalouomega committed:
Adding different linear solvers

loumalouomega committed:
Update test

loumalouomega committed:
Use diagonal norm

loumalouomega committed:
Update script

loumalouomega committed:
Custom variable

loumalouomega committed:
Custom variable

loumalouomega committed:
Independent method

loumalouomega committed:
Update methods and tests

loumalouomega committed:
Extend test

loumalouomega committed:
Adding more methods

loumalouomega committed:
Adding AUXILIAR_CONSTRAINT_SCALE_FACTOR

loumalouomega committed:
Adding AUXILIAR_CONSTRAINT_SCALE_FACTOR

loumalouomega committed:
Consistent with paper

loumalouomega committed:
Update test

loumalouomega committed:
Set to zero 

This is essential for iterative solvers

loumalouomega committed:
Update solver to AMGCL

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Using numerical limit

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Update test

loumalouomega commented:
BTW This may interest to ROM developers @Rbravo555

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
Ping

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Update B&S

loumalouomega committed:
Update utility

loumalouomega committed:
Update solver

loumalouomega committed:
Update solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
Ping @KratosMultiphysics/technical-committee

loumalouomega committed:
Merge master

loumalouomega committed:
Removing arguments constructor

loumalouomega committed:
Exposing parameters constructor

loumalouomega committed:
Update tests

loumalouomega committed:
Correct solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Update solver

loumalouomega committed:
Update

loumalouomega committed:
Expose to python

loumalouomega committed:
Update solver

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -110,7 +110,11 @@ def GetDefaultSettings(cls):
             "compute_reactions": true,
-            "block_builder": true,
+            "block_builder" : true,
+            "consider_lagrange_multiplier_constraint_resolution" : "None",
```
```suggestion
            "consider_lagrange_multiplier_constraint_resolution" : "none",
```

BTW those are very specialized settings, I think we should consider making the B&S settings a sub-parameter

loumalouomega replied:
If you see https://github.com/KratosMultiphysics/Kratos/pull/3185, you will see that I separate the configuration for everything. I will comment there

philbucher replied:
so you are planning to do it altogether?
Why not start now if you already know how it will be?
Otherwise it will be more difficult to adapt

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -425,7 +429,18 @@ def _create_linear_solver(self):
         linear_solver = self.get_linear_solver()
         if self.settings["block_builder"].GetBool():
-            builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolver(linear_solver)
+            bs_params = KratosMultiphysics.Parameters("{}")
```
basically you are doing this here

philbucher commented on applications/StructuralMechanicsApplication/tests/structural_mechanics_test_factory.py:
```diff
@@ -38,6 +38,9 @@ def setUp(self):
                     }""")
                 ProjectParameters["solver_settings"].AddValue("linear_solver_settings", default_lin_solver_settings)
+            if not ProjectParameters["solver_settings"].Has("silent_warnings_block_builder_and_solver"):
+                silent_warnings_block_builder_and_solver = KratosMultiphysics.Parameters("""{"silent_warnings_block_builder_and_solver" : true}""")
```
@RiccardoRossi do you agree with this?

loumalouomega replied:
This is in #5905, better to comment there

philbucher reviewed (changes requested):
blocking to avoid accidental merge, this contains multiple other PRs and affects many files

loumalouomega committed:
none

loumalouomega committed:
Using vector

loumalouomega committed:
Proper OMP loop

loumalouomega committed:
OMp min/max/etc

loumalouomega committed:
Revert

loumalouomega committed:
Update B&S accordingly

loumalouomega committed:
Update solver

loumalouomega committed:
Update test

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
First somplification

loumalouomega committed:
Removing variable added

loumalouomega committed:
Removing variable

loumalouomega committed:
Update

loumalouomega committed:
Windows fix

loumalouomega committed:
Windows fix

loumalouomega committed:
Windows fix

loumalouomega committed:
Fix warning ⚠

loumalouomega committed:
Fix compilation

loumalouomega committed:
Update test

loumalouomega committed:
Update solver

loumalouomega committed:
Minor

loumalouomega committed:
Suggestion (I)

loumalouomega committed:
Suggestion (II)

loumalouomega committed:
Suggestion (III)

loumalouomega committed:
Update

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Clean up

loumalouomega committed:
Minor refactor

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Update

loumalouomega committed:
@RiccardoRossi suggestion

loumalouomega committed:
Revert "@RiccardoRossi suggestion"

This reverts commit 08a879e5630ea2b63b6da475a254701a60e98416.

loumalouomega committed:
Conflict

loumalouomega committed:
no_scaling

loumalouomega committed:
@RiccardoRossi suggestions

loumalouomega committed:
Missing const

loumalouomega committed:
First private

loumalouomega committed:
Avoiding accessing data twice

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Remove from python

loumalouomega committed:
Minor

loumalouomega committed:
Corrections

loumalouomega committed:
Update test

loumalouomega committed:
Typos

loumalouomega committed:
Update

loumalouomega committed:
Corrections

loumalouomega committed:
Update mScaleFactor

loumalouomega committed:
Update test

loumalouomega committed:
KRATOS_ERROR_IF

loumalouomega committed:
Adapt ApplyDirichletConditions

loumalouomega committed:
Update test

loumalouomega committed:
Update solver

loumalouomega committed:
static_cast<int>(system_size)

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -433,7 +440,15 @@ def _create_linear_solver(self):
+            bs_params = self.settings["builder_and_solver_settings"]
+            consider_lagrange_multiplier_constraint_resolution = self.settings["builder_and_solver_settings"]["consider_lagrange_multiplier_constraint_resolution"].GetString()
+            if consider_lagrange_multiplier_constraint_resolution == "none":
+                bs_params.RemoveValue("constraint_scale_factor")
```
IMHO it is really bad practice to remove stuff from the input parameters.

loumalouomega committed:
@RiccardoROssi and @philbutcher suggestions

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

philbucher committed:
formatting to dismiss review

loumalouomega committed:
Adapting changes

loumalouomega committed:
Revert "Removing variable"

This reverts commit d16d0ea708bf85e6f775051b456e3ddcaea31f24.

loumalouomega committed:
Revert "Removing variable added"

This reverts commit 33fd8ae623f7e7d5fdbf4f2f86e361830e071e4f.

loumalouomega committed:
Revert assigment

loumalouomega committed:
Update test

loumalouomega committed:
Fix I

loumalouomega committed:
Fix II

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Conflict

loumalouomega committed:
Update B&S

loumalouomega committed:
Using enum

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Toward update

loumalouomega committed:
Using enums also here

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
This is ready @KratosMultiphysics/technical-committee after #5905

loumalouomega commented:
Well requires #5919

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Not needed

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
This is ready for review @KratosMultiphysics/technical-committee and @KratosMultiphysics/implementation-committee

philbucher commented:
why are there changes in the BlockBuilder? :sweat_smile:

loumalouomega commented:
> why are there changes in the BlockBuilder? sweat_smile

There should not be..., let me check

loumalouomega commented:
> why are there changes in the BlockBuilder? sweat_smile

Merge conflicts, I need a moment

loumalouomega commented:
> > why are there changes in the BlockBuilder? sweat_smile
> 
> Merge conflicts, I need a moment

OK, these methods must be protected, my fault

loumalouomega commented:
> > > why are there changes in the BlockBuilder? sweat_smile
> > 
> > 
> > Merge conflicts, I need a moment
> 
> OK, these methods must be protected, my fault

(not private)

philbucher commented on applications/StructuralMechanicsApplication/tests/structural_mechanics_test_factory.py:
```diff
@@ -49,6 +49,17 @@ def setUp(self):
 
             SelectAndVerifyLinearSolver(ProjectParameters, self.skipTest)
 
+            solver_type = ProjectParameters["solver_settings"]["linear_solver_settings"]["solver_type"].GetString()
```
merge conflict?

loumalouomega replied:
Nope, we said to follow that structure

philbucher replied:
I am 99,9% sure this is a merge conflict, check a few lines above!

loumalouomega replied:
okay

loumalouomega replied:
OK, probably, I will check later

philbucher reviewed (commented):
so you only moved the methods from private to protected?

loumalouomega committed:
Conflict

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
Now is ready

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+    {
+        KRATOS_TRY;
+
+        BaseType::SetUpDofSet(pScheme, rModelPart);
```
please remove the methods that only call the baseclass

RiccardoRossi replied:
+1

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+
+            // Free memory
+            b_lm.resize(0, false);
+            b_modified.resize(0, false);
```
same as in the previous PR, can be removed as it anyway goes out of scope here

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+
+            // Copy the RHS
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(BaseType::mEquationSystemSize); ++i) {
```
why not using the `Space` function?

loumalouomega replied:
True

loumalouomega replied:
In fact if you check the sizes, this is not coping or setting the whole vector

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+            // Fill with zeros
+            #pragma omp parallel for
+            for (int i = static_cast<int>(BaseType::mEquationSystemSize); i < static_cast<int>(total_size_of_system); ++i) {
+                b_modified[i] = 0.0;
```
same, isn't there a function in the `Space` for this?

loumalouomega replied:
Check that is not setting to zero the whole vector

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+
+            // Free memory
+            b_lm.resize(0, false);
+            b_modified.resize(0, false);
```
idem

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+            // Free memory
+            copy_of_A.resize(0, 0, false);
+            copy_of_T.resize(0, 0, false);
+            transpose_of_T.resize(0, 0, false);
```
idem

philbucher reviewed (commented):
some initial comments

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -115,8 +115,11 @@ def GetDefaultSettings(cls):
-                "silent_warnings"                    : false
+                "consider_lagrange_multiplier_constraint_resolution" : "none",
+                "diagonal_values_for_dirichlet_dofs"                 : "use_max_diagonal",
+                "constraint_scale_factor"                            : "use_mean_diagonal",
```
these settings are not there for the "normal" builder and solver...so we cannot use them by default

RiccardoRossi replied:
? this is important...

philbucher replied:
this should not be included here IMO but directly be given to the B&S, where it knows its parameters and can do the validation

loumalouomega replied:
OK; I will see how can I do to refactor without breaking anything (the problem are as always the checks)

RiccardoRossi commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -438,7 +441,14 @@ def _create_builder_and_solver(self):
             bs_params = self.settings["builder_and_solver_settings"]
-            builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolver(linear_solver, bs_params)
+            consider_lagrange_multiplier_constraint_resolution = self.settings["builder_and_solver_settings"]["consider_lagrange_multiplier_constraint_resolution"].GetString()
+            if consider_lagrange_multiplier_constraint_resolution == "none":
```
i really think it is a bad strategy to remove settings... 

in any case the validation is not using them (that's exactly why we shouldn't do recursive validation). They should be validated by the builder_and_solver itself, at least as soon as factories will be merged

loumalouomega replied:
I know, but the factories are not yet merged...

RiccardoRossi commented on kratos/python/add_containers_to_python.cpp:
```diff
@@ -320,6 +320,8 @@ void  AddContainersToPython(pybind11::module& m)
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, RESIDUAL_NORM )
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, CONVERGENCE_RATIO )
     KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, BUILD_SCALE_FACTOR )
+    KRATOS_REGISTER_IN_PYTHON_VARIABLE(m, CONSTRAINT_SCALE_FACTOR )
```
why don't we use the same as the build_scale_factor? is there a good reason for which we need fine control over those?

loumalouomega replied:
The paper separates the tow types, de facto I consider the same value but may be different

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+     * @param rA The LHS matrix
+     * @param rb The RHS vector
+     */
+    void Build(
```
how is this different from the base class?

loumalouomega replied:
It is adding the LM contributions

RiccardoRossi replied:
i mean, here you call BaseType for everything relevant... it is identical to the baseclass

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+        KRATOS_TRY
+
+        // Compute the norm
+        const double norm_b = (TSparseSpace::Size(rb) != 0) ? TSparseSpace::TwoNorm(rb) : 0.0;
```
this also looks the same as the base class no?

RiccardoRossi replied:
ping

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+     * @param rb The RHS vector
+     * @param rModelPart The model part of the problem to solve
+     */
+    void SystemSolveWithPhysics(
```
does not need to be reimplemented...

loumalouomega replied:
If it is not calling the base class it does

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+        KRATOS_TRY
+
+        // Resize to the proper size
+        const SizeType total_system_size = (BaseType::mOptions.Is(DOUBLE_LAGRANGE_MULTIPLIER)) ? BaseType::mEquationSystemSize + 2 * BaseType::mSlaveIds.size() : BaseType::mEquationSystemSize + BaseType::mSlaveIds.size();
```
i don't understand this.  Lagrange multipliers are in the dofset, so they should be present in the mEquationSystemSize no?

what is the size of A.size1 or rDx?

loumalouomega replied:
LM are not in the dofset, are not dof, are complementary dofs only existing in this B&S, at the end these are saved as reactions directly in the nodes

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1234 @@
+            }
+
+            // If not previously computed we compute
+            if (TSparseSpace::TwoNorm(BaseType::mT) < std::numeric_limits<double>::epsilon()) {
```
it is expensive to compute the norm of T. Why do you need to do it this way??

loumalouomega replied:
I can add a flag to know if the mT has been already computed

loumalouomega replied:
I forgot about this, one sec

loumalouomega replied:
Done

RiccardoRossi commented on kratos/tests/test_linear_constraints.py:
```diff
@@ -19,98 +21,90 @@ def setUp(self):
+                "silent_warnings"                                    : false,
+                "consider_lagrange_multiplier_constraint_resolution" : "Single"
+            }""")
+            self.builder_and_solver = KM.ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier(self.linear_solver, params)
```
i would put this in a different test. It is better to have many small test files than few large ones

loumalouomega replied:
This is the best way to avoid code duplication, it only changes the B&S

RiccardoRossi replied:
please do split it. My point is that the current version of  it is much easier to understand than after your modification. In short i definitely would prefer two tests i understand that one that i don't...

loumalouomega replied:
In https://github.com/KratosMultiphysics/Kratos/pull/6789 I separate the modification of the test. The quantity of code added to the test **is only 4 lines after that is merged**. Reducing code duplication is one of the best way to avoid errors, we see that constantly each time we do an important change in the API and we refactor code and we see like the same error was done several times

loumalouomega committed:
Remocing methods calling base class

loumalouomega committed:
Out of scope

loumalouomega committed:
Revert "Using space methods"

This reverts commit ad6fcb28c5518bb8f9bfb7ba1214e75e5898d232.

loumalouomega committed:
Using space methods

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Skip remove warnings

loumalouomega commented:
@RiccardoRossi and @philbucher, comments applied, @RiccardoRossi there are some questions that I replied you directly

loumalouomega committed:
Using flag

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
> @RiccardoRossi and @philbucher, comments applied, @RiccardoRossi there are some questions that I replied you directly

Is it everything OK?

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1182 @@
+     * @param rDx The Unknowns vector
+     * @param rb The RHS vector
+     */
+    void ApplyDirichletConditions(
```
i also think this function is identical to the base class

loumalouomega replied:
At some point, it wasn't, removing

loumalouomega replied:
Nope, they are different, otherwise tests fail. The difference is:

~~~cpp
// Filling with ones the LM dofs
        #pragma omp parallel for firstprivate(ndofs)
        for (int k = ndofs; k<static_cast<int>(system_size); ++k) {
            scaling_factors[k] = 1.0;
        }
~~~

loumalouomega committed:
Codacy suggestion: Trailing whitespace

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1182 @@
+    double mConstraintFactor = 0.0;                                    /// The constraint scale factor
+    double mAuxiliarConstraintFactor = 0.0;                            /// The auxiliar constraint scale factor
+
+    CONSTRAINT_FACTOR mConstraintFactorConsidered;                  /// The value considered for the constraint factor
```
why the type is in capitals?

loumalouomega replied:
It is a enum, enums are with capitals

loumalouomega replied:
Or not?

RiccardoRossi replied:
as i understand, the variable name should be capitals (as a variable to understand), but not the type

RiccardoRossi commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1182 @@
+     * @brief This method constructs the master slaeve constraint structure
+     * @param rModelPart The problem model part
+     */
+    void ConstructMasterSlaveConstraintsStructure(ModelPart& rModelPart) override
```
is this changed wrt the base class?

loumalouomega replied:
Completely, is a completely different matrix

loumalouomega committed:
Typo

loumalouomega commented:
@RiccardoRossi I cannot use SystemSolve from teh base class, as there is an operation not compatible with the LM B&S. Basically 

~~~cpp
if(mT.size1() != 0) //if there are master-slave constraints
        {
            //recover solution of the original problem
            TSystemVectorType Dxmodified = Dx;

            TSparseSpace::Mult(mT, Dxmodified, Dx);
        }

~~~

loumalouomega commented:
Dear @RiccardoRossi the Build method is different, because after apply the constraints the size of A is different, and if I call the base Build, it will fail. i already check, all the methods that are modified,a re modified because a reason

RiccardoRossi commented:
ok, i will review again tomorrow but in principle i am ok with your
comments.

the only thing i am missing is how thst is used. do you need to use a
special choice of variabkes?

On Wed, Apr 15, 2020, 8:45 PM Vicente Mataix Ferrándiz <
notifications@github.com> wrote:

> Dear @RiccardoRossi <https://github.com/RiccardoRossi> the Build method
> is different, because after apply the constraints the size of A is
> different, and if I call the base Build, it will fail. i already check, all
> the methods that are modified,a re modified because a reason
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5939#issuecomment-614213868>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5PWEPT3226OF7AJ72ZLK3RMX6GFANCNFSM4JOMCQQQ>
> .
>

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
[Core] Adding advanced setting into the block B&S

This gives more flexibility

loumalouomega committed:
Update

loumalouomega committed:
Update B&S

loumalouomega committed:
Update solver

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -115,11 +115,13 @@ def GetDefaultSettings(cls):
+                "advanced_settings"                                  : {
+                    "consider_lagrange_multiplier_constraint_resolution" : "none",
+                    "constraint_scale_factor"                            : "use_mean_diagonal",
+                    "auxiliar_constraint_scale_factor"                   : "use_mean_diagonal"
```
But those are still weong for e.g. the elimB&S
I think you should just pass them on to the B&S for usage there

loumalouomega replied:
The elimination B&S is not currently constructed with Parameters, but in case we add one we should add an empty "advanced_settings"

philbucher replied:
I don’t think the manual manipulation is the way to go here

loumalouomega replied:
Precisely this is the only way to avoid manual manipulation

philbucher replied:
I don’t think so, the cleanest solution would be to just pass the parameters of the B&S to the B&S

loumalouomega replied:
In that case the B&S with the same parameters will crash in the Parameters validation

philbucher replied:
I propose the following:
~~~js
{
    ...
    "builder_and_solver_settings" : {
    // settings that are used by the structural-solver
    "advanced_builder_and_solver_settings" : { // maybe find a better name for this
        // this will be directly passed to the CTor of the B&S
    }
}
~~~

loumalouomega replied:
Ok, but, what do I pass to the CTor if the validation may crash in some B&S?, do I call them directly "block_builder_and_solver_settings" and "lm_builder_and_solver_settings"?

philbucher replied:
I would only Pass the params to the B&S that have this Ctor
And in the future to all of them

loumalouomega replied:
My question is, ho do you make this parameters compatible with the constructor of other B&S, can you write an example?

philbucher replied:
In here those settings would be empty, exactly the same how we do it with the linear solver
This is possible bcs we don’t do recursive validation

philbucher replied:
Thats why I proposed the splitting into settings that go into the b&S and settings that are used by the solver

loumalouomega replied:
But that's basically what I have done with the advance_settings

loumalouomega replied:
Ok, I will update today later

philbucher replied:
I took a look again and I want to make it more clear:
this is how it would look in the solver (names of the settings are just examples):
~~~json
"builder_and_solver_settings" : {
    "use_block_builder" : true,
    "use_lagrange_B&S" : false,
    "advanced_settings"                                  : {
        "diagonal_values_for_dirichlet_dofs"                 : "use_max_diagonal",
        "silent_warnings"                                    : false,
        "consider_lagrange_multiplier_constraint_resolution" : "none",
        "constraint_scale_factor"                            : "use_mean_diagonal",
        "auxiliar_constraint_scale_factor"                   : "use_mean_diagonal"
    }
}
~~~

loumalouomega replied:
If I move the "use_block_builder" this will change current behavior, I would do that in a future PR

loumalouomega committed:
Update C++ test

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:
```diff
@@ -137,7 +137,8 @@ class ResidualBasedBlockBuilderAndSolver
             "diagonal_values_for_dirichlet_dofs" : "use_max_diagonal",
-            "silent_warnings"                    : false
+            "silent_warnings"                    : false,
+            "advanced_settings"                  : {}
```
the point is to NOT have this here, only in the solver!

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -438,7 +443,11 @@ def _create_builder_and_solver(self):
+            if consider_lagrange_multiplier_constraint_resolution == "none":
+                builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolver(linear_solver, bs_params)
+            else:
+                builder_and_solver = KratosMultiphysics.ResidualBasedBlockBuilderAndSolverWithLagrangeMultiplier(linear_solver, bs_params)
```
the `bs_params` and directly the `advanced_settings` and without any manipulation

philbucher reviewed (commented):
I put some more comments to make it more clear

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
[Core] Adding mthods that validate but skip the subparameters to Parameters

This way an error is avoided when we only want to validate one level of Parameters

loumalouomega committed:
Refactor (I)

loumalouomega committed:
Refactor (II)

loumalouomega committed:
Refactor (III)

loumalouomega committed:
Update solvers

loumalouomega committed:
Update test

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Missing

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
This depends on #6822

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

adityaghantasala commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1184 @@
+                contribution_coefficients(2, 2) = -mAuxiliarConstraintFactor;
+
+                // Assemble the matrix (NOTE: Like the identity matrix is created inside the condition must be used meanwhile is alive, so inside the condition)
+                SparseMatrixMultiplicationUtility::AssembleSparseMatrixByBlocks(rA, matrices_p_blocks, contribution_coefficients, transpose_blocks);
```
the `MultiplicationUtility` is doing assembly ? why don't you put it into the `space` ? It is misleading. By the wax does this work in MPI ?

loumalouomega replied:
Yes, is doing the assembly. No, it only works in OpenMP

loumalouomega replied:
It is not part of the space because includes many non standard things...

adityaghantasala replied:
ok .. then the naming and/or including an assembly method in MultiplicationUtility is not consistent.

loumalouomega replied:
This method is already merged with the master, do you want me to modify the name that already exists and break its use in other implementations?

adityaghantasala replied:
I saw it used only in tests .. so I asked. If you say it will break other impl. Its fine ... !

loumalouomega replied:
Only if this is supported by the @KratosMultiphysics/technical-committee , what name do you propose? The method basically is assembling a new sparse matrix from blocks

adityaghantasala commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_lagrange_multiplier.h:
```diff
@@ -0,0 +1,1184 @@
+            BuildMasterSlaveConstraints(rModelPart);
+
+            // Copy the LHS to avoid memory errors
+            TSystemMatrixType copy_of_A(rA);
```
Can't you avoid making those copies. These copies consume a lot of memory as the number of DOFs increase. 
In this function all of A, b and T have a copy. This momentarily will cause memory spike which could be a problem.

loumalouomega replied:
It cannot be avoided, the rA is a reference that we are going to expand into the new form. In order to build a sparse matrix, you need to build the structure, so I cannot preserve the old values, I must copy them

loumalouomega replied:
In any case, you don't see it, but the block builder and solver does this, but it does using the SparseMultiplicationUtilities, so you cannot see it directly

loumalouomega replied:
For the constraints BTW

adityaghantasala replied:
Well all I am saying this will cause memory spikes .. and using this will be restrictive in terms of the size of the problem very quickly. 
Imagine having a A of size x GB (without constraints) , when using this, even if there a very few constraints, (not considering T) one will have a memory of minimum 2x GB of memory for a short while and in the case it does not fit in the memory this is going to cause a crash.

loumalouomega replied:
I understand, but there is no way to avoid this IMO due to the rigid behaviour of the sparse matrix design

loumalouomega replied:
@adityaghantasala I found a way to avoid the copy, create an empty sparse matrix and do a swap between them

loumalouomega replied:
Problem solved

philbucher commented:
I see that I still have to review this in terms of handling the settings
Will do it in the next days, sorry for the delay

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega commented:
> I see that I still have to review this in terms of handling the settings
> Will do it in the next days, sorry for the delay

Ping?

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Minor

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Doing a swap instead of a copy

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

Commit:
Adding Name method

Commit:
Adding GetDefaultParameters method

Commit:
Adding AssignSettings

Commit:
Refactor method

Commit:
Renaming category

Commit:
Update test

Commit:
Update test

Commit:
Update settings

Commit:
Adding to Block empty parameters

Commit:
Minor

Commit:
Missing comma

Commit:
Minor

Commit:
Missing import

Commit:
Missing recursive validation

Commit:
Missing start

philbucher commented on kratos/includes/kratos_parameters.h:
```diff
@@ -815,7 +815,15 @@ class KRATOS_API(KRATOS_CORE) Parameters
     void ValidateAndAssignDefaults(const Parameters& rDefaultParameters);
```
please revert the changes in the Parameters, those should no longer be necessary

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -105,10 +116,18 @@ def GetDefaultSettings(cls):
+                "advanced_settings"                      : {
+                    "diagonal_values_for_dirichlet_dofs" : "use_max_diagonal",
+                    "silent_warnings"                    : false,
+                    "advanced_builder_and_solver_settings"                   : {
```
Now that we have #7290 this should no longer be necessary

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -127,6 +146,15 @@ def GetDefaultSettings(cls):
+        """This function validates the settings of the solver
+        """
+        default_settings = self.GetDefaultSettings()
+        self.settings.ValidateAndAssignDefaults(default_settings)
```
```suggestion
        super().ValidateSettings()
```

philbucher commented on kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:
```diff
@@ -1224,10 +1224,11 @@ class ResidualBasedBlockBuilderAndSolver
+            "block_builder"                        : true,
+            "diagonal_values_for_dirichlet_dofs"   : "use_max_diagonal",
+            "silent_warnings"                      : false,
+            "advanced_builder_and_solver_settings" : {}
```
```suggestion
```

not necessary due to #7290

loumalouomega replied:
The parameters must be at least exist, or the validation will crash

philbucher replied:
I dont understand, can you explain why/where?

loumalouomega replied:
When you use the method ValidateAndAssignSettings you need to have all the parameters from the parameters to be checked defined in the base class, or will crash when passing throw the base class. As it is not  RecursivelyValidateAndAssignSettings, just the name will be enough, that's why I defined the parameters as "advanced_builder_and_solver_settings", so can be adapted to any derived classes

philbucher replied:
but the derived classes should add the additional parameters themselves so this doesn't happen. This is exactly what #7290 does

loumalouomega replied:
Yes it does, but requires at least to have the base parameters defined, otherwise will crash. See https://github.com/KratosMultiphysics/Kratos/blob/6ab06bdfb4ee206895a050fa349429b9ff2b2542/kratos/solving_strategies/strategies/residualbased_linear_strategy.h#L786, this was agreed in #7290

philbucher replied:
but this is the Strategy not the B&S

loumalouomega replied:
It is exactly the same concept, this is due to the Parameters behavior, not from the B&S

loumalouomega replied:
If I removed, it simply won't work (that's why it stopped compiling after I solved the merge conflicts this weekend)

philbucher replied:
Ok then I will try myself

philbucher replied:
compiling...

philbucher replied:
now should work

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -55,6 +56,16 @@ def __init__(self, model, custom_settings):
 
+        settings_have_use_block_builder = custom_settings.Has("block_builder")
+
+        if settings_have_use_block_builder:
+            kratos_utilities.IssueDeprecationWarning('MechanicalSolver', 'Using "block_builder", please move it to "builder_and_solver_settings" as "use_block_builder"')
+            if not custom_settings.Has("builder_and_solver_settings"):
+                custom_settings.AddEmptyValue("builder_and_solver_settings")
+
+            custom_settings["builder_and_solver_settings"].AddValue("use_block_builder", custom_settings["block_builder"])
+            custom_settings.RemoveValue("block_builder")
```
:+1:

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:
```diff
@@ -105,10 +116,18 @@ def GetDefaultSettings(cls):
-                "silent_warnings"                    : false
+                "use_block_builder" : true,
+                "use_lagrange_BS"   : false,
+                "advanced_settings"                      : {
```
for the record, I find this a good idea of separating global settings (e.g. `use_block_builder`) from detailed/advanced settings (like `silent_warnings`)
=> lets also use this for the other classes!

philbucher reviewed (commented):
minor comments related to cleanup otherwise this is ready!

loumalouomega committed:
Suggestion

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

philbucher committed:
fixing parameters issue

philbucher committed:
fix test

loumalouomega committed:
Merge branch 'master' into core/doble-lm-builder-and-solver

loumalouomega committed:
Merge remote-tracking branch 'origin/core/doble-lm-builder-and-solver' into core/doble-lm-builder-and-solver

loumalouomega committed:
Revert parameters

philbucher committed:
revert changes in test

philbucher reviewed (approved):
:tada:
